### DISCUSSION TITLE
RFC-0003 20-08-2024 TDD (Test Drive Development)

### DISCUSSION THREAD

**Discussion by Leopgf** [Avila-Tek/fullstack-template#2] [Ideas]
# Contexto

Este RFC viene de la necesidad comenzar a escribir test automatizado en el backend para atacar los fallos y errores de la implementación de lógica de backend de manera preventiva y no de manera reactiva. Esta ocurriendo en varios proyectos que no se logran identificar errores de flujos importantes de backend sino hasta mucho tiempo después de un deploy o cuando un cliente en producción reporta que esta teniendo un fallo. Entendemos que no es factible estar probando constantemente las mismas funcionalidades por razones de tiempo y de ineficiencia. 
Otro punto importante a destacar es que se tiene poco criterio para identificar diferentes casos bordes en funcionalidades importante para la aplicación, por lo tanto, el API responde bien para flujo principal, pero esta completamente roto para flujos secundarios.

# Alcance

Debemos encontrar la mejor manera para introducir pruebas automatizadas en el back para nuestros proyectos y tomar en consideración el impacto en el rendimiento como % de tareas completadas con pruebas y el % de tareas con bugs en producción 

# Propuesta

Estamos considerando trabajar bajo el esquema de TDD, en donde le vamos a exigir a los desarrolladores que escriban primero pruebas unitarias antes de escribir el código de una funcionalidad que satisface un requerimiento.
Debemos considerar los siguientes pasos para lograrlo:

1. Definir el alcance de las pruebas
2. Capacitar a los devs para escribir pruebas 
3. Como vamos a validar que se esta siguiendo correctamente el estilo de TDD
4. Definir como se va incluir en la planificación de las HUs
5. Configurar los actions del build para que ejecute las pruebas
6. Seguir los principios KISS y YAGNI
7. Prueba de concepto con un proyecto piloto

# Detalles técnicos 

Para lograr esto debemos 

1. Acordar estructura de carpetas para los archivos de test
2. Incluir back dependencias para hacer testing 
3. Implementar configuraciones en nuestros CI/CD

**Comment by Zoomelectrico** [comment]
Entiendo la propuesta! Creo que deberíamos enfocarnos más en ver que podemos hacer para ir haciendo el rollout poco a poco.

Por ejemplo en #1 ya tenemos la resuelto modificar el template para incluir las pruebas en la parte de backend, claramente necesitamos hacer un trabajo de formación.

Creo que deberíamos enfocarnos en cosas como: Cuales son los entrenamientos que tenemos que realizar para lograr cumplir con el objetivo de hacer pruebas unitarias


PD: De estos principios es que esta hablando @Leopgf 
![image](https://github.com/user-attachments/assets/b26f49ad-3fea-44b4-9bc3-8b7923012b83)

**Reply by Leopgf** [reply]
Mi propuesta inicial para ir haciendo rollout poco a poco es 

1. Capacitación de los devs 
Tenemos que recordarles a los que ya sabían e indicarle a los nuevos recursos donde se encuentra disponible el curso de jest proporcionado por Avila. Para aligerar la carga del curso y hacer más realista el objetivo de completarlo debemos destacar que hay unos módulos que son para front, otros para back y otros para ambos.
2. Validar que se este siguiendo correctamente el estilo de TDD
Aquí mi propuesta sería indicar a los recursos que primero deben subir un PR con todas las pruebas en red y luego otro PR con el código refactorizado con todas las pruebas en green 
3. Definir como se va incluir en el plan de trabajo 
Mi propuesta inicial sería crear HUs o tareas separadas para aquellas funciones a las que se le va a aplicar TDD debido a que incialmente toma más tiempo trabajar bajo este estilo y los beneficios se ven a mediano-largo plazo
4. Prueba de concepto 
Proponer de la misma manera como se hizo con la integración de Drone CI uno o varios proyectos piloto para evaluar el impacto. Considerar que dichas pruebas pueden tomar como mínimo un mes antes de poder empezar a ver los resultados

**Reply by andrespd99** [reply]
> 2\. Aquí mi propuesta sería indicar a los recursos que primero deben subir un PR con todas las pruebas en red y luego otro PR con el código refactorizado con todas las pruebas en green

Esto lo veo poco factible, porque estarías haciendo un PR de código incompleto. Cuando sigues TDD muchas veces haces referencia a funciones que no has definido, variables que aun no existen, etc. Yo veo el TDD no tanto en el proceso de entrega, si no en el mindset de programación y de resolución de problemas.

**Reply by Lstanislao** [reply]
Para mi que he investigado sobre este tema el TDD es cambiar la mentalidad con la que la gente programa. Para eso te puedes enfocar que la gente debe de entender como escribir codigo testeable. En mi caso tengo pendiente ver y si es bueno compartir con el equipo algunas secciones de este curso donde hablan de como se debe escribir el codigo para ser testaeado: Writing Testable Code https://frontendmasters.com/courses/web-app-testing/ Ojo no se que tecnologia usan en el curso yo hablo es mas de los conceptos

**Reply by Lstanislao** [reply]
Del mismo modo que Andres comenta deberia ser un solo PR queda como responsabilidad del Dev cumplir con el TDD

**Comment by andrespd99** [comment]
Estoy de acuerdo con seguir TDD en los proyectos de back. Como esto es un cambio de paradigma completamente tenemos que estar claros de que va a tomar un tiempo para que el equipo se adapte y lo siga correctamente. 

Lo primordial sería volver expertos en TDD a los miembros clave del equipo (leads, Sr., etc.) y generar una capacitación en cascada–que esas personas se encarguen de capacitar a los miembros mas Jr. del equipo.

Ya cuando TDD sea pan de cada día, será cuestión de que en el proceso de onboarding de los nuevos miembros del equipo se les enseñe TDD correctamente.

**Reply by Zoomelectrico** [reply]
Estoy de acuerdo con lo que dice @andrespd99, es importante empezar por agregar pruebas a la cultura de programación antes de volvernos loco con TDD, para complementar también con la respuesta @Leopgf es importante redefinir el punto 2

**Reply by andrespd99** [reply]
Igualmente, lo primordial es que se comience a testear. TDD es simplemente un estilo de pensar y programar, lo que hace un poco dificil de evaluar si se está siguiendo o no.

De todas formas es un estilo que cuando la gente lo entienda y vea los beneficios va a ser natural que lo apliquen. Quizás el enfoque ahorita debería ser en capacitar a los devs en testing y obligar a escribir pruebas unitarias **relevantes, acertadas y suficientes**.

**Reply by Lstanislao** [reply]
Estoy de acuerdo por eso creo que los primeros pasos es lo que estamos haciendo de la estructura para pruebas y si podemos ir introduciendo conceptops de testeable code en el equipo. Que sea un cambio progresivo porque si bien el TDD suena muy bonito y lo es, es un cambio significativo y estamos llecando multiples cambios actualmente

**Reply by Leopgf** [reply]
Entiendo sus comentarios y estoy de acuerdo. Lo tomo en cuenta para redefinir la propuesta inicial y comenzar a capacitar en cascada a los miembros del backend según su nivel de seniority para escribir pruebas unitarias con especial enfoque para aquellos proyectos que ya se encuentren en producción

**Comment by Zoomelectrico** [comment]
He estado leyendo bastante al respecto de este tema y como puede ser el mejor mecanismo para afrontar.

Lo primero es que tenemos que invertir tiempo en formación y entrenamiento en pruebas. Para ello, lo que considero importante es que los devs se ven los cursos de pruebas. La otra actividad que veo importante es que hagamos una serie de talleres para construir una funcionalidad sencilla a nivel de backend y que podamos hacer pruebas a esa funcionalidad. 

Luego, lo siguiente que veo importante es dar una guía y/o una serie de parámetros a considerar para escribir pruebas unitarias.

Y lo tercero es hacer el seguimiento a la implementación de estas pruebas, con estos puntos creo que podemos lograr la meta de que las pruebas sean parte importante de nuestro proceso de desarrollo

**Reply by Zoomelectrico** [reply]
Para profundizar más en el segundo punto, creo importante que nuestras pruebas enfoque en representar las reglas de negocio de las funcionalidades que estamos probando. Por ejemplo, supongamos que vamos probar una funcionalidad en Spacars.

## feat: seleccionar productos para un servicios

Para esta funcionalidad tenemos que considerar lo siguiente. Dado un modelo de vehículo con su año la función debe regresar un producto de tipo filtro de aceite, un producto de tipo aceite y un producto de tipo servicio. En caso de que el filtro de aceite y/o el aceite no este disponible en el inventario deberá dar un error, para avisarle al usuario que no es posible realizar una reservación.

Dadas estas condiciones iniciales se puede suponer que al menos las siguientes pruebas deben hacerse.

1. Dado un modelo de carro y un tipo de aceite, se encuentran todos los productos necesarios.
2. Dado un modelo de carro y un tipo de aceite, no se encuentra aceites de ese tipo.
3. Dado un modelo de carro y un tipo de aceite, no se encuentra filtro de aceite para ese modelo.
4. Dado un modelo de carro y un tipo de aceite, no se encuentra ese modelo de carro.
5. Dado un modelo de carro, el servicio rechaza la petición.
6. Dado un tipo de aceite, el servicio rechaza la petición.

Con esos 6 casos de prueban se pueden garantizar que la cobertura del código es lo suficientemente relevante para representar las reglas de negocio de la aplicación.

Luego podemos discutir a nivel más especifico que tips a nivel de código se pueden utilizar para escribir estas pruebas, como los que se mencionan en algunos de estos link:

1. [https://leanylabs.com/blog/good-unit-tests/](https://leanylabs.com/blog/good-unit-tests/)
2. [https://medium.com/@kaanfurkanc/unit-testing-best-practices-3a8b0ddd88b5](https://medium.com/@kaanfurkanc/unit-testing-best-practices-3a8b0ddd88b5)

**Reply by Lstanislao** [reply]
Si mas o menos este fue el enfoque que lleve en front con las pruebas, es criar la mentalidad de unit test !

**Reply by lesanpi** [reply]
En mi opinión la mejor forma de introducir TDD es demostrando a los devs la necesidad que se tiene de utilizar los tests y el beneficio que se tendrán. 
Para mi, se me hace más realista comenzar a hacer TDD en los flujos principales y core de las reglas del negocio, ya que esas son los casos fundamentales que necesitas cubrir con los tests, lo que hará que el developer sienta que es valioso hacer dichos tests. En la estimación considero que debería ser así:
1. Identificar un feature core de la aplicación, lo suficientemente importante, en el que la aplicación no sirve sin ese flujo.
2. Agregar una segunda task de TDD dentro de la HU, aqui se pueden ir identificando pruebas importantes. 
3. Iniciar la definición de las pruebas.
5. Desarrollar el feature. (obviamente deberían pasar todas las pruebas)
6. Hacer PR.


Mención importante:
Ya que estamos teniendo mucha dedicación a las pruebas dentro de la empresa. Tanto las pruebas E2E, como las pruebas unitarias de backend y frontend, tenemos que tener cuidado en que no deben solaparse, en el sentido que estemos probando los mismos casos.

**Comment by Leopgf** [comment]
# Conclusión 

Actualmente debemos fomentar la cultura en el departamento de backend para realizar pruebas de unit testing antes de empezar a utilizar TDD como estilo de programación, por lo tanto, se inicio con la capacitación de los miembros de departamento de backend para que tengan todas las herramientas y conocimientos necesarios para empezar a escribir pruebas unitarias en el código del API. Es importante agregar que se contara con un base en lark para que los devs de back puedan hacer tracking de su progreso en el curso de testing.

El objetivo a cumplir es escribir pruebas en el API, específicamente pruebas unitarias que permitan validar el funcionamiento de las funciones más pequeñas que no tengan conexión con base de datos, ni con integraciones. Esto último quiere decir que no se van a realizar pruebas integración a funciones de servicios que realicen varios llamados a base de datos o otros servicios (Otras APIs). Estos son los pasos a seguir para realizar la integración de las pruebas en el código de back de la misma manera como se integró con el front. 

Se harán funciones a manera de explicación que sirvan de ejemplo para que los devs puedan guiarse para escribir sus pruebas, esto en el WIKI de onboarding de backend. De la misma manera, se construirá un repositorio con funcionalidades básicas a manera de recurso para que los devs puedan correr las pruebas y ver más de cerca el funcionamiento de las mismas.  

Finalmente, se incluirá un tiempo adicional a las historias de usuario para que los devs puedan escribir y entregar sus pruebas unitarias
